### PR TITLE
feat: add types for svg-baker-runtime

### PR DESCRIPTION
**Pull request by vagusX**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by vagusX** [comment]
@peterblazejewicz could you please help review this pr?

File: types/svg-baker-runtime/src/browser-sprite.d.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import Sprite from './sprite';
+import BrowserSpriteSymbol from './browser-symbol';
+
+import SpriteConfig from './sprite.config';
+import BrowserSpriteConfig from './browser-sprite.config';
+
+export default class BrowserSprite extends Sprite {
+    constructor(cfg?: SpriteConfig & BrowserSpriteConfig);
+
+    config: SpriteConfig & BrowserSpriteConfig;
```
Comment (peterblazejewicz) [inline types/svg-baker-runtime/src/browser-sprite.d.ts:10]: this should be on base class, either `protected` or `public` (default).
Are you sure about intersection type here?
https://github.com/JetBrains/svg-mixer/blob/68f4f77f780d64883e26719a882c9893888c7099/packages/svg-baker-runtime/src/sprite.js#L10

File: types/svg-baker-runtime/src/browser-sprite.d.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    node: Element | null;
```
Comment (peterblazejewicz) [inline types/svg-baker-runtime/src/browser-sprite.d.ts:11]: this this one supposed to be public? It's access via 'root' getters or queries for state with 'isMounted', but is this used directly somewhere? (breaking encapsulation intent)

File: types/svg-baker-runtime/src/sprite.config.d.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+export default interface SpriteConfig {
```
Comment (peterblazejewicz) [inline types/svg-baker-runtime/src/sprite.config.d.ts:1]:
```suggestion
export interface SpriteConfig {
```

File: types/svg-baker-runtime/src/sprite.d.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import SpriteSymbol from './symbol';
+import SpriteConfig from './sprite.config';
+
+export default class Sprite {
+    config: SpriteConfig;
+    symbols: SpriteSymbol[];
+
+    constructor(config: SpriteConfig);
```
Comment (peterblazejewicz) [inline types/svg-baker-runtime/src/sprite.d.ts:8]:
```suggestion
    constructor(config?: SpriteConfig);
```
it always defaults to default configuration:
https://github.com/JetBrains/svg-mixer/blob/68f4f77f780d64883e26719a882c9893888c7099/packages/svg-baker-runtime/src/sprite.js#L10
For some reason the module avoids using optional params defaults

File: types/svg-baker-runtime/src/sprite.d.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    /**
+     * Remove symbol & destroy it
+     * @return `true` - symbol was found & successfully destroyed, `false` - otherwise
+     */
+    remove: (id: string) => boolean;
+    find: (id: string) => SpriteSymbol | null;
+    has: (id: string) => boolean;
+    stringify: () => string;
+    toString: () => string;
+    destroy: () => void;
```
Comment (peterblazejewicz) [inline types/svg-baker-runtime/src/sprite.d.ts:25]:
```suggestion
```
this one is probably protected/private or at least non documented (like in other places)
https://github.com/JetBrains/svg-mixer/blob/68f4f77f780d64883e26719a882c9893888c7099/packages/svg-baker-runtime/src/sprite.js#L82
https://github.com/JetBrains/svg-mixer/blob/68f4f77f780d64883e26719a882c9893888c7099/packages/svg-baker-runtime/src/sprite.js#L82

File: types/svg-baker-runtime/src/symbol.d.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+export default class SpriteSymbol {
+    id: string;
+    viewBox: string;
+    content: string;
```
Comment (peterblazejewicz) [inline types/svg-baker-runtime/src/symbol.d.ts:4]: are those public? (used when consuming library)

File: types/svg-baker-runtime/src/browser-sprite.d.ts (hunk 1 above)
Comment (vagusX) [reply types/svg-baker-runtime/src/browser-sprite.d.ts:10]: SuperClass `Sprite` already has the type `config` and it was typed as `SpriteConfig`, this subclass would handle more config options by combining superclass's config and itself config.
You can see this https://github.com/JetBrains/svg-mixer/blob/68f4f77f780d64883e26719a882c9893888c7099/packages/svg-baker-runtime/src/browser-sprite.js#L37, subclass's config contains own `autoConfigure` field and  would be consumed by `this._autoConfigure` https://github.com/JetBrains/svg-mixer/blob/68f4f77f780d64883e26719a882c9893888c7099/packages/svg-baker-runtime/src/browser-sprite.js#L93

File: types/svg-baker-runtime/src/symbol.d.ts (hunk 6 above)
Comment (vagusX) [reply types/svg-baker-runtime/src/symbol.d.ts:4]: yes

File: types/svg-baker-runtime/src/browser-sprite.d.ts (hunk 2 above)
Comment (vagusX) [reply types/svg-baker-runtime/src/browser-sprite.d.ts:11]: In my opinion, these properties/methods starts with underscore should be private and others should be public or protected. Like these samples: 
* `_autoConfigure`: https://github.com/JetBrains/svg-mixer/blob/v1/packages/svg-baker-runtime/src/browser-sprite.js#L93
* `https://github.com/JetBrains/svg-mixer/blob/v1/packages/svg-baker-runtime/src/browser-sprite.js#L116` https://github.com/JetBrains/svg-mixer/blob/v1/packages/svg-baker-runtime/src/browser-sprite.js#L116
* `_emitter` https://github.com/JetBrains/svg-mixer/blob/v1/packages/svg-baker-runtime/src/browser-sprite.js#L32

And I can see some `@private` jsdoc markers in the source code, like [this search result in sourcegraph](
https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/JetBrains/svg-mixer%24%40v1+%40private&patternType=literal)

image

So I would like to change this private marked properties/methods to private in d.ts, and others would following the reviews and the source codes themselves.

Comment (peterblazejewicz) [reply types/svg-baker-runtime/src/browser-sprite.d.ts:11]: OK, so I would say `destron` is protected (or @internal to lib), there is no mention of any kind of destroy calls on docs site:
https://github.com/JetBrains/svg-mixer/tree/master/packages/svg-mixer

My point here is only, if the declarations files are not autogenerated, there is no need to provide private/internal symbols, the module is to be consumed by client code.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@vagusX thx!
fix the CI test, just run:
- prettier:
```
npm run prettier -- --write types/svg-baker-runtime/**/*.ts.
```
- run local test
```
npm test svg-baker-runtime
```

**Comment by vagusX** [comment]
dependent on this package: [https://www.npmjs.com/browse/depended/svg-baker-runtime](https://www.npmjs.com/browse/depended/svg-baker-runtime)

| packages |  used methods/propreties |
| --- | --- |
| [https://github.com/JetBrains/svg-sprite-loader/blob/051b82e48db26e1bb234ab9ba5f3ea9fdbc3b9db/runtime/browser-sprite.js#L14](https://github.com/JetBrains/svg-sprite-loader/blob/051b82e48db26e1bb234ab9ba5f3ea9fdbc3b9db/runtime/browser-sprite.js#L14) | constructored with attrs |
| [https://github.com/sveltevk/icons/blob/9f98df197da218d9fe2ba4385ac286cfcd2e69aa/tasks/sprite.ts](https://github.com/sveltevk/icons/blob/9f98df197da218d9fe2ba4385ac286cfcd2e69aa/tasks/sprite.ts) | add/attach/mount  |
| [https://github.com/boltdesignsystem/bolt/blob/master/packages/components/bolt-icon/src/icon.standalone.js](https://github.com/boltdesignsystem/bolt/blob/master/packages/components/bolt-icon/src/icon.standalone.js) | node/attach/mount |
| [https://github.com/VKCOM/icons/blob/5815717e46c39e6d31216aac0adf16358c41454d/src/sprite.ts](https://github.com/VKCOM/icons/blob/5815717e46c39e6d31216aac0adf16358c41454d/src/sprite.ts) | add/attach/mount  |

**Comment by vagusX** [comment]
> review

Updated

**Comment by peterblazejewicz** [comment]
it's just enough to add 'protected' to THE fields you want to limit access to.[Everything is public by default](https://www.typescriptlang.org/docs/handbook/2/classes.html#public), no linting config changes required:
```diff
2015170d06 wip
diff --git a/types/svg-baker-runtime/src/browser-sprite.d.ts b/types/svg-baker-runtime/src/browser-sprite.d.ts
index c161444cff..756c983b8c 100644
--- a/types/svg-baker-runtime/src/browser-sprite.d.ts
+++ b/types/svg-baker-runtime/src/browser-sprite.d.ts
@@ -11,8 +11,8 @@ export default class BrowserSprite extends Sprite {
      * override
      */
     protected config: SpriteConfig & BrowserSpriteConfig;
-    public node: Element | null;
-    public isMounted: boolean;
+    node: Element | null;
+    isMounted: boolean;
 
     /**
      * override
@@ -21,30 +21,30 @@ export default class BrowserSprite extends Sprite {
      * @fires Events#SYMBOL_MOUNT
      * @return `true` - symbol was added, `false` - replaced
      */
-    public add(symbol: BrowserSpriteSymbol): boolean;
+    add(symbol: BrowserSpriteSymbol): boolean;
 
     /**
      * Attach to existing DOM node
      * @return attached DOM Element. null if node to attach not found.
      */
-    public attach(target: string | Element): Element | null;
+    attach(target: string | Element): Element | null;
 
     /**
      * @fires Events#MOUNT
      * @return rendered sprite node. null if mount node not found.
      */
-    public mount(target?: string, prepend?: boolean): Element | null;
+    mount(target?: string, prepend?: boolean): Element | null;
 
-    public render(): Element;
+    render(): Element;
 
     /**
      * Detach sprite from the DOM
      */
-    public unmount(): void;
+    unmount(): void;
 
     /**
      * Update URLs in sprite and usage elements
      * @return `true` - URLs was updated, `false` - sprite is not mounted
      */
-    public updateUrls(oldUrl: string, newUrl: string): boolean;
+    updateUrls(oldUrl: string, newUrl: string): boolean;
 }
diff --git a/types/svg-baker-runtime/src/browser-symbol.d.ts b/types/svg-baker-runtime/src/browser-symbol.d.ts
index fd7add366e..de9f4521b9 100644
--- a/types/svg-baker-runtime/src/browser-symbol.d.ts
+++ b/types/svg-baker-runtime/src/browser-symbol.d.ts
@@ -1,16 +1,16 @@
 import SpriteSymbol from './symbol';
 
 export default class BrowserSpriteSymbol extends SpriteSymbol {
-    public static createFromExistingNode(node: Element): BrowserSpriteSymbol;
+    static createFromExistingNode(node: Element): BrowserSpriteSymbol;
 
-    public isMounted: boolean;
+    isMounted: boolean;
 
-    public mount: (target: Element | string) => Element;
-    public render: () => Element;
-    public unmount: () => void;
+    mount: (target: Element | string) => Element;
+    render: () => Element;
+    unmount: () => void;
 
     /**
      * override
      */
-    public destroy: () => void;
+    destroy: () => void;
 }
diff --git a/types/svg-baker-runtime/src/sprite.d.ts b/types/svg-baker-runtime/src/sprite.d.ts
index e56fffb4c0..3615e25462 100644
--- a/types/svg-baker-runtime/src/sprite.d.ts
+++ b/types/svg-baker-runtime/src/sprite.d.ts
@@ -11,16 +11,16 @@ export default class Sprite {
      * Add new symbol. If symbol with the same id exists it will be replaced.
      * @return `true` - symbol was added, `false` - replaced
      */
-    public add(symbol: SpriteSymbol): boolean;
+    add(symbol: SpriteSymbol): boolean;
 
     /**
      * Remove symbol & destroy it
      * @return `true` - symbol was found & successfully destroyed, `false` - otherwise
      */
-    public remove: (id: string) => boolean;
-    public find: (id: string) => SpriteSymbol | null;
-    public has: (id: string) => boolean;
-    public stringify: () => string;
-    public toString: () => string;
-    public destroy: () => void;
+    remove: (id: string) => boolean;
+    find: (id: string) => SpriteSymbol | null;
+    has: (id: string) => boolean;
+    stringify: () => string;
+    toString: () => string;
+    destroy: () => void;
 }
diff --git a/types/svg-baker-runtime/src/symbol.d.ts b/types/svg-baker-runtime/src/symbol.d.ts
index 8b77c0f416..93e5a9d73c 100644
--- a/types/svg-baker-runtime/src/symbol.d.ts
+++ b/types/svg-baker-runtime/src/symbol.d.ts
@@ -1,11 +1,11 @@
 export default class SpriteSymbol {
-    public id: string;
-    public viewBox: string;
-    public content: string;
+    id: string;
+    viewBox: string;
+    content: string;
 
     constructor(cfg: { id: string; viewBox: string; content: string });
 
-    public stringify: () => string;
-    public toString: () => string;
-    public destroy: () => void;
+    stringify: () => string;
+    toString: () => string;
+    destroy: () => void;
 }
diff --git a/types/svg-baker-runtime/tslint.json b/types/svg-baker-runtime/tslint.json
index 73ba332809..2efa283002 100644
--- a/types/svg-baker-runtime/tslint.json
+++ b/types/svg-baker-runtime/tslint.json
@@ -1,6 +1,3 @@
 {
-    "extends": "@definitelytyped/dtslint/dt.json",
-    "rules": {
-        "member-access": false
-    }
+    "extends": "@definitelytyped/dtslint/dt.json"
 }
```
otherwise it's ready to go

**Comment by vagusX** [comment]
> it's just enough to add 'protected' to THE fields you want to limit access to.[Everything is public by default](https://www.typescriptlang.org/docs/handbook/2/classes.html#public), no linting config changes required:
> 
> ```diff
> 2015170d06 wip
> diff --git a/types/svg-baker-runtime/src/browser-sprite.d.ts b/types/svg-baker-runtime/src/browser-sprite.d.ts
> index c161444cff..756c983b8c 100644
> --- a/types/svg-baker-runtime/src/browser-sprite.d.ts
> +++ b/types/svg-baker-runtime/src/browser-sprite.d.ts
> @@ -11,8 +11,8 @@ export default class BrowserSprite extends Sprite {
>       * override
>       */
>      protected config: SpriteConfig & BrowserSpriteConfig;
> -    public node: Element | null;
> -    public isMounted: boolean;
> +    node: Element | null;
> +    isMounted: boolean;
>  
>      /**
>       * override
> @@ -21,30 +21,30 @@ export default class BrowserSprite extends Sprite {
>       * @fires Events#SYMBOL_MOUNT
>       * @return `true` - symbol was added, `false` - replaced
>       */
> -    public add(symbol: BrowserSpriteSymbol): boolean;
> +    add(symbol: BrowserSpriteSymbol): boolean;
>  
>      /**
>       * Attach to existing DOM node
>       * @return attached DOM Element. null if node to attach not found.
>       */
> -    public attach(target: string | Element): Element | null;
> +    attach(target: string | Element): Element | null;
>  
>      /**
>       * @fires Events#MOUNT
>       * @return rendered sprite node. null if mount node not found.
>       */
> -    public mount(target?: string, prepend?: boolean): Element | null;
> +    mount(target?: string, prepend?: boolean): Element | null;
>  
> -    public render(): Element;
> +    render(): Element;
>  
>      /**
>       * Detach sprite from the DOM
>       */
> -    public unmount(): void;
> +    unmount(): void;
>  
>      /**
>       * Update URLs in sprite and usage elements
>       * @return `true` - URLs was updated, `false` - sprite is not mounted
>       */
> -    public updateUrls(oldUrl: string, newUrl: string): boolean;
> +    updateUrls(oldUrl: string, newUrl: string): boolean;
>  }
> diff --git a/types/svg-baker-runtime/src/browser-symbol.d.ts b/types/svg-baker-runtime/src/browser-symbol.d.ts
> index fd7add366e..de9f4521b9 100644
> --- a/types/svg-baker-runtime/src/browser-symbol.d.ts
> +++ b/types/svg-baker-runtime/src/browser-symbol.d.ts
> @@ -1,16 +1,16 @@
>  import SpriteSymbol from './symbol';
>  
>  export default class BrowserSpriteSymbol extends SpriteSymbol {
> -    public static createFromExistingNode(node: Element): BrowserSpriteSymbol;
> +    static createFromExistingNode(node: Element): BrowserSpriteSymbol;
>  
> -    public isMounted: boolean;
> +    isMounted: boolean;
>  
> -    public mount: (target: Element | string) => Element;
> -    public render: () => Element;
> -    public unmount: () => void;
> +    mount: (target: Element | string) => Element;
> +    render: () => Element;
> +    unmount: () => void;
>  
>      /**
>       * override
>       */
> -    public destroy: () => void;
> +    destroy: () => void;
>  }
> diff --git a/types/svg-baker-runtime/src/sprite.d.ts b/types/svg-baker-runtime/src/sprite.d.ts
> index e56fffb4c0..3615e25462 100644
> --- a/types/svg-baker-runtime/src/sprite.d.ts
> +++ b/types/svg-baker-runtime/src/sprite.d.ts
> @@ -11,16 +11,16 @@ export default class Sprite {
>       * Add new symbol. If symbol with the same id exists it will be replaced.
>       * @return `true` - symbol was added, `false` - replaced
>       */
> -    public add(symbol: SpriteSymbol): boolean;
> +    add(symbol: SpriteSymbol): boolean;
>  
>      /**
>       * Remove symbol & destroy it
>       * @return `true` - symbol was found & successfully destroyed, `false` - otherwise
>       */
> -    public remove: (id: string) => boolean;
> -    public find: (id: string) => SpriteSymbol | null;
> -    public has: (id: string) => boolean;
> -    public stringify: () => string;
> -    public toString: () => string;
> -    public destroy: () => void;
> +    remove: (id: string) => boolean;
> +    find: (id: string) => SpriteSymbol | null;
> +    has: (id: string) => boolean;
> +    stringify: () => string;
> +    toString: () => string;
> +    destroy: () => void;
>  }
> diff --git a/types/svg-baker-runtime/src/symbol.d.ts b/types/svg-baker-runtime/src/symbol.d.ts
> index 8b77c0f416..93e5a9d73c 100644
> --- a/types/svg-baker-runtime/src/symbol.d.ts
> +++ b/types/svg-baker-runtime/src/symbol.d.ts
> @@ -1,11 +1,11 @@
>  export default class SpriteSymbol {
> -    public id: string;
> -    public viewBox: string;
> -    public content: string;
> +    id: string;
> +    viewBox: string;
> +    content: string;
>  
>      constructor(cfg: { id: string; viewBox: string; content: string });
>  
> -    public stringify: () => string;
> -    public toString: () => string;
> -    public destroy: () => void;
> +    stringify: () => string;
> +    toString: () => string;
> +    destroy: () => void;
>  }
> diff --git a/types/svg-baker-runtime/tslint.json b/types/svg-baker-runtime/tslint.json
> index 73ba332809..2efa283002 100644
> --- a/types/svg-baker-runtime/tslint.json
> +++ b/types/svg-baker-runtime/tslint.json
> @@ -1,6 +1,3 @@
>  {
> -    "extends": "@definitelytyped/dtslint/dt.json",
> -    "rules": {
> -        "member-access": false
> -    }
> +    "extends": "@definitelytyped/dtslint/dt.json"
>  }
> ```
> 
> otherwise it's ready to go

updated

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@vagusX thx!
Thx for your patience!

**Comment by vagusX** [comment]
@peterblazejewicz thanks for your reviews

**Comment by vagusX** [comment]
Ready to merge @typescript-bot

**Pull request merged**